cpp/src_prims/selection/knn.h

```diff
@@ -104,93 +170,451 @@ void merge_tables(long n, long k, long nshard, float *distances, long *labels,
+   *        to correspond to n_parts.
+   * @param n_int_streams size of internalStreams. When this is <= 0, only the
+   *        user stream will be used.
+   * @param translations translation ids for indices when index rows represent
+   *        non-contiguous partitions
    */
-template <typename IntType = int>
-void brute_force_knn(float **input, int *sizes, int n_params, IntType D,
-                     float *search_items, IntType n, long *res_I, float *res_D,
-                     IntType k, cudaStream_t s) {
-  std::vector<long> *id_ranges = new std::vector<long>();
+template <typename IntType = int,
+          Distance::DistanceType DistanceType = Distance::EucUnexpandedL2>
+void brute_force_knn(std::vector<float *> &input, std::vector<int> &sizes,
+                     IntType D, float *search_items, IntType n, int64_t *res_I,
+                     float *res_D, IntType k,
+                     std::shared_ptr<deviceAllocator> allocator,
+                     cudaStream_t userStream,
+                     cudaStream_t *internalStreams = nullptr,
+                     int n_int_streams = 0, bool rowMajorIndex = true,
+                     bool rowMajorQuery = true,
+                     std::vector<int64_t> *translations = nullptr) {
+  ASSERT(DistanceType == Distance::EucUnexpandedL2 ||
+           DistanceType == Distance::EucUnexpandedL2Sqrt,
+         "Only EucUnexpandedL2Sqrt and EucUnexpandedL2 metrics are supported "
+         "currently.");
+
+  ASSERT(input.size() == sizes.size(),
+         "input and sizes vectors should be the same size");
+
+  int n_params = input.size();
+
+  std::vector<int64_t> *id_ranges;
+  if (translations == nullptr) {
+    // If we don't have explicit translations
+    // for offsets of the indices, build them
+    // from the local partitions
+    id_ranges = new std::vector<int64_t>();
+    int64_t total_n = 0;
+    for (int i = 0; i < n_params; i++) {
+      if (i < input.size()) {
+        id_ranges->push_back(total_n);
+      }
+      total_n += sizes[i];
+    }
+  } else {
+    // otherwise, use the given translations
+    id_ranges = translations;
+  }
 
-  IntType total_n = 0;
+  int device;
+  CUDA_CHECK(cudaGetDevice(&device));
 
-  for (int i = 0; i < n_params; i++) {
-    if (i < n_params)  // if i < sizes[i]
-      id_ranges->push_back(total_n);
-    total_n += sizes[i];
-  }
+  device_buffer<int64_t> trans(allocator, userStream, id_ranges->size());
+  updateDevice(trans.data(), id_ranges->data(), id_ranges->size(), userStream);
 
-  float *result_D = new float[k * size_t(n)];
-  long *result_I = new long[k * size_t(n)];
+  device_buffer<float> all_D(allocator, userStream, input.size() * k * n);
+  device_buffer<int64_t> all_I(allocator, userStream, input.size() * k * n);
 
-  float *all_D = new float[n_params * k * size_t(n)];
-  long *all_I = new long[n_params * k * size_t(n)];
+  if (n_int_streams > 0) CUDA_CHECK(cudaStreamSynchronize(userStream));
```

Comment:
this sync should be present irrespective of `n_int_streams`, no?

Reply:
If only the `userStream` is being used, I believe the synchronize is unnecessary here because the KNN query from the different partitions will also be performed in the `userStream`. I only synchronize when the internal streams from the handle are being used, since those could end up reading memory which has not been written by the `userStream` (assuming this could be a call in a larger pipeline of primitives).